include/s3select.h

```diff
@@ -505,12 +529,22 @@ struct s3select : public bsc::grammar<s3select>
 
       arithmetic_argument = (float_number)[BOOST_BIND_ACTION(push_float_number)] |  (number)[BOOST_BIND_ACTION(push_number)] | (column_pos)[BOOST_BIND_ACTION(push_column_pos)] |
                             (string)[BOOST_BIND_ACTION(push_string)] |
-                            (cast) |
+                            (cast) | (substr) |
                             (function) | (variable)[BOOST_BIND_ACTION(push_variable)] ;//function is pushed by right-term
 
       cast = (bsc::str_p("cast") >> '(' >> arithmetic_expression >> bsc::str_p("as") >> (data_type)[BOOST_BIND_ACTION(push_data_type)] >> ')') [BOOST_BIND_ACTION(push_cast_expr)];
 
       data_type = (bsc::str_p("int") | bsc::str_p("float") | bsc::str_p("string") |  bsc::str_p("timestamp") );
+     
+      substr = (substr_one_arg) | (substr_two_arg) | (substr_from) | (substr_from_for);
+
+      substr_one_arg = (bsc::str_p("substring") >> '(' >> (arithmetic_expression >> bsc::str_p(",") >> arithmetic_expression) >> ')') [BOOST_BIND_ACTION(push_substr_one_arg)];
+
+      substr_two_arg = (bsc::str_p("substring") >> '(' >> (arithmetic_expression >> bsc::str_p(",") >> arithmetic_expression >> bsc::str_p(",") >> arithmetic_expression) >> ')') [BOOST_BIND_ACTION(push_substr_two_arg)];
```

Comment:
no need for substr_one_arg / substr_two_arg rules.
the rule for function accept both rules , the operator will handle those different functions.

Reply:
Argument order reverses for  push_substr_from and push_substr_from_for in operator. I have modified substring operator to accommodate that. Should I create new operator for 'from' and 'from_for'?

Reply:
it better to do it in single operator (re-use code).
main consideration is performance.
the /operator()/ is very intensive , thus code should be compact as possible.